src/mpid/ch4/shm/ipc/xpmem/globals.c

```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpidimpl.h"
+#include "xpmem_pre.h"
+
+MPIDI_IPC_xpmem_global_t MPIDI_IPC_xpmem_global;
+
+#ifdef MPL_USE_DBG_LOGGING
+MPL_dbg_class MPIDI_CH4_SHM_IPC_XPMEM_GENERAL;
+#endif
+
+/* Preallocated segment objects */
+MPIDI_IPC_xpmem_seg_t MPIDI_IPC_xpmem_seg_mem_direct[MPIDI_IPC_XPMEM_SEG_PREALLOC];
```

Comment:
Why do we need preallocated segments?

Reply:
These are used to add cache entries in the critical path, it seems. Might be the reason. I haven't written this code so maybe @KaimingOuyang can provide a better response?

Reply:
We have preallocated objects for requests to optimize a pointer deference for small messages.  That seems completely unnecessary for IPC handles, which are not even created for small messages.

Reply:
For XPMEM, the object will be created for every 4KB message if it hasn't been cached.

Reply:
Why would it not be cached?

Reply:
(removed my previous comment to avoid digressing)

What I mean is: it's OK to preallocate objects at INIT time.  But do these need to be static objects, rather than in a pool?  In other words, the direct/indirect distinction is an overkill for IPC handles.

Reply:
I should mention in my previous comment that if the buffer is first time used, it would not be cached. For example, if the program sends 1000 4KB messages each uses a different sbuf, the `malloc` will be called 1000 times if we remove the preallocated objects for segment.

Reply:
Yeah, I know what you are saying.  I probably used bad terminology earlier.  I agree that preallocated objects are needed.  I'm only wondering if we need the "direct" objects.  That's messy, doesn't allow for multiple pools, and so on.  Do those give any benefit for IPC handles?

Reply:
We used it only because this has already been implemented as common routine in MPI. We did not know the `MPIDIU_create_buf_pool` function when writing the XPMEM code. You are right that the static part is not beneficial here.

Since this PR should focus only on refactoring the shmmod structure, I'd move the memory pool fix into a separate PR. Does it make sense?

Reply:
> Since this PR should focus only on refactoring the shmmod structure, I'd move the memory pool fix into a separate PR. Does it make sense?

Yeah, I agree.